src/mrinufft/trajectories/densities.py

```diff
@@ -0,0 +1,134 @@
+"""Sampling densities and methods."""
+
+import numpy as np
+import numpy.fft as nf
+import numpy.linalg as nl
+import numpy.random as nr
+import pywt as pw
+from sklearn.cluster import BisectingKMeans, KMeans
+from tqdm.auto import tqdm
+
+from .utils import KMAX
+
+
+def sample_from_density(nb_samples, density, method="random"):
+    rng = nr.default_rng()
+
+    shape = density.shape
+    nb_dims = len(shape)
+    max_nb_samples = np.prod(shape)
+
+    if nb_samples > max_nb_samples:
+        raise ValueError("`nb_samples` must be lower than the size of `density`.")
+
+    if method == "random":
+        choices = rng.choice(
+            np.arange(max_nb_samples),
+            size=nb_samples,
+            p=density.flatten(),
+            replace=False,
+        )
+        locations = np.indices(shape).reshape((nb_dims, -1))[:, choices]
+        locations = locations.T
+        locations = 2 * KMAX * locations / np.max(locations) - KMAX
+    elif method == "lloyd":
+        kmeans = (
+            KMeans(n_clusters=nb_samples)
+            if nb_dims <= 2
+            else BisectingKMeans(n_clusters=nb_samples)
+        )
+        kmeans.fit(
+            np.indices(density.shape).reshape((nb_dims, -1)).T,
+            sample_weight=density.flatten(),
+        )
+        locations = kmeans.cluster_centers_ - np.array(density.shape) / 2
+        locations = KMAX * locations / np.max(np.abs(locations))
+    else:
+        raise ValueError(f"Unknown sampling method {method}.")
+    return locations
+
+
+def create_cutoff_decay_density(shape, cutoff, decay, resolution=None):
```

Comment:
How about a more generic density defined in Eq. 10 in https://onlinelibrary.wiley.com/doi/full/10.1002/mrm.29702 ? 
Its a very short update.
Also, having support for elliptical sampling enabled or not can be helpful (we soon plan to disable it :P )

Reply:
It is based on that equation actually. What difference do you see here ?

Reply:
I am confused, this density is already based on that equation. What difference do you see ?
And what do you mean by elliptical sampling ? Anisotropic ? Why do you plan to disable it and from where?

Reply:
Anisotropic and more importantly elliptical sampling, a way to enable or disable mask.

Reply:
Somehow that dodges all of my questions xD Could you reformulate pls ?

Reply:
We need to support elliptical or non elliptical density

We also need to support anisotropy, but I think that we already do?

Reply:
I had to fix it but I think it can now handle anisotropy and elliptical densities